File: plugins/tokenmonitor/server/js/test/usbids.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import {
  deviceTable,
  classifyVIDPID,
  labelFor,
  TIER_PROBE,
  TIER_SHARED,
  TIER_REGISTRY_MATCH,
} from "../src/usbprov/usbids.js";

const here = dirname(fileURLToPath(import.meta.url));
function findCompat(rel) {
  let dir = here;
  for (let i = 0; i < 12; i++) {
    const c = join(dir, "compat", rel);
    if (existsSync(c)) return c;
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return null;
}

const fixPath = findCompat("usb-ids.json");
const skip = fixPath ? false : "compat/usb-ids.json unavailable (standalone checkout)";
const fixture = fixPath ? JSON.parse(readFileSync(fixPath, "utf8")) : { devices: [] };

// The hardcoded deviceTable must be byte-for-value identical to
// compat/usb-ids.json — same VID/PID (parsed), tier and label, same order and
// count. This is what makes "hardcoded, not loaded at runtime" honest.
test("hardcoded deviceTable matches usb-ids.json fixture", { skip }, () => {
  assert.equal(deviceTable.length, fixture.devices.length, "device count");
  for (let i = 0; i < fixture.devices.length; i++) {
    const fd = fixture.devices[i];
    const vid = parseInt(fd.vid, 16);
    const pid = parseInt(fd.pid, 16);
    assert.ok(Number.isInteger(vid) && vid >= 0 && vid <= 0xffff, `fixture[${i}] vid`);
    assert.ok(Number.isInteger(pid) && pid >= 0 && pid <= 0xffff, `fixture[${i}] pid`);
    const hc = deviceTable[i];
    assert.equal(hc.vid, vid, `entry ${i} vid`);
    assert.equal(hc.pid, pid, `entry ${i} pid`);
    assert.equal(hc.tier, fd.tier, `entry ${i} tier`);
    assert.equal(hc.label, fd.label, `entry ${i} label`);
  }
});

test("no duplicate (vid,pid) entries", () => {
  const seen = new Map();
  for (let i = 0; i < deviceTable.length; i++) {
    const e = deviceTable[i];
    const key = (e.vid << 16) | e.pid;
    assert.equal(seen.has(key), false, `duplicate at entry ${i}`);
    seen.set(key, i);
  }
});

test("every hardcoded tier is probe or shared (never registry-match)", () => {
  for (const e of deviceTable) {
    assert.ok(e.tier === TIER_PROBE || e.tier === TIER_SHARED, `${e.vid.toString(16)} tier ${e.tier}`);
    assert.notEqual(e.tier, TIER_REGISTRY_MATCH);
  }
});

test("classifyVIDPID", () => {
  assert.deepEqual(classifyVIDPID(0x303a, 0x1001), { tier: TIER_PROBE, found: true });
  assert.deepEqual(classifyVIDPID(0x1a86, 0x7523), { tier: TIER_SHARED, found: true });
  // Unknown degrades to the most restrictive tier and reports found=false.
  assert.deepEqual(classifyVIDPID(0xdead, 0xbeef), { tier: TIER_SHARED, found: false });
});

test("labelFor", () => {
  assert.equal(labelFor(0x303a, 0x1001), "Espressif USB-Serial/JTAG (ESP32-S3 / C3 / C6)");
  assert.equal(labelFor(0xdead, 0xbeef), "");
});

File: plugins/tokenmonitor/server/go/internal/usbprov/usbids_test.go
package usbprov

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"testing"
)

// findUSBIDs walks up to the repo-root compat/usb-ids.json — the spec + test
// fixture. Like the framing vectors it is NOT vendored into server/compat/
// (it is never read at runtime; each runtime hardcodes its own copy), so this
// walks all the way to the monorepo root. Skips on a standalone checkout.
func findUSBIDs(t *testing.T) string {
	t.Helper()
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}
	dir := wd
	for i := 0; i < 9; i++ {
		candidate := filepath.Join(dir, "compat", "usb-ids.json")
		if _, err := os.Stat(candidate); err == nil {
			return candidate
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	t.Skipf("compat/usb-ids.json not found upward from %s (standalone checkout)", wd)
	return ""
}

type usbIDsDoc struct {
	Devices []struct {
		VID   string `json:"vid"`
		PID   string `json:"pid"`
		Tier  string `json:"tier"`
		Label string `json:"label"`
	} `json:"devices"`
}

// TestUSBIDs_MatchFixture asserts the hardcoded deviceTable is byte-for-value
// identical to compat/usb-ids.json — same VID/PID (parsed), tier and label, in
// the same order and with the same count. This is what makes "hardcoded, not
// loaded at runtime" honest: an unreviewed change to the fixture (or a
// corrupted vendoring) that widened the set of writable devices would fail
// here rather than silently taking effect.
func TestUSBIDs_MatchFixture(t *testing.T) {
	raw, err := os.ReadFile(findUSBIDs(t))
	if err != nil {
		t.Fatalf("read usb-ids.json: %v", err)
	}
	var doc usbIDsDoc
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("parse usb-ids.json: %v", err)
	}

	if len(doc.Devices) != len(deviceTable) {
		t.Fatalf("device count: hardcoded %d, fixture %d", len(deviceTable), len(doc.Devices))
	}
	for i, fd := range doc.Devices {
		vid, err := strconv.ParseUint(fd.VID, 0, 16)
		if err != nil {
			t.Errorf("fixture[%d] vid %q not 16-bit hex: %v", i, fd.VID, err)
			continue
		}
		pid, err := strconv.ParseUint(fd.PID, 0, 16)
		if err != nil {
			t.Errorf("fixture[%d] pid %q not 16-bit hex: %v", i, fd.PID, err)
			continue
		}
		hc := deviceTable[i]
		if uint16(vid) != hc.VID || uint16(pid) != hc.PID {
			t.Errorf("entry %d id: hardcoded %04x:%04x, fixture %04x:%04x",
				i, hc.VID, hc.PID, vid, pid)
		}
		if string(hc.Tier) != fd.Tier {
			t.Errorf("entry %d (%04x:%04x) tier: hardcoded %q, fixture %q",
				i, hc.VID, hc.PID, hc.Tier, fd.Tier)
		}
		if hc.Label != fd.Label {
			t.Errorf("entry %d (%04x:%04x) label:\n  hardcoded %q\n  fixture   %q",
				i, hc.VID, hc.PID, hc.Label, fd.Label)
		}
	}
}

// TestUSBIDs_NoDuplicates enforces the fixture's "duplicate_or_contradictory
// fail closed" rule on the hardcoded table: two entries for the same (vid,pid)
// are contradictory and must never ship.
func TestUSBIDs_NoDuplicates(t *testing.T) {
	seen := map[uint32]int{}
	for i, e := range deviceTable {
		key := uint32(e.VID)<<16 | uint32(e.PID)
		if j, dup := seen[key]; dup {
			t.Errorf("duplicate (vid,pid) %04x:%04x at entries %d and %d", e.VID, e.PID, j, i)
		}
		seen[key] = i
	}
}

// TestUSBIDs_KnownTiers enforces the fixture's "unknown_tier degrades to
// shared" spirit on the hardcoded table: every hardcoded tier must be one of
// the two VID/PID-keyed tiers (registry-match is never in the table).
func TestUSBIDs_KnownTiers(t *testing.T) {
	for _, e := range deviceTable {
		switch e.Tier {
		case TierProbe, TierShared:
			// ok
		case TierRegistryMatch:
			t.Errorf("%04x:%04x is registry-match — that tier is runtime-resolved, never in the table", e.VID, e.PID)
		default:
			t.Errorf("%04x:%04x has unknown tier %q", e.VID, e.PID, e.Tier)
		}
	}
}

func TestClassifyVIDPID(t *testing.T) {
	if tier, found := ClassifyVIDPID(0x303a, 0x1001); !found || tier != TierProbe {
		t.Errorf("Espressif should classify as probe/found, got %q/%v", tier, found)
	}
	if tier, found := ClassifyVIDPID(0x1a86, 0x7523); !found || tier != TierShared {
		t.Errorf("CH340 should classify as shared/found, got %q/%v", tier, found)
	}
	// An unknown serial device degrades to the most restrictive tier and
	// reports found=false.
	if tier, found := ClassifyVIDPID(0xdead, 0xbeef); tier != TierShared || found {
		t.Errorf("unknown device must be shared/not-found, got %q/%v", tier, found)
	}
}
